File: 22.generate-parentheses.py
class Solution:
    def add(self, ret, s, n, m):
        if n > m:
            return
        if m == 0 and n == 0:
            ret.append(s)
        if n > 0:
            self.add(ret, s + '(', n - 1, m)
        if m > 0:
            self.add(ret, s + ')', n, m - 1)

    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        ret = []
        self.add(ret, "", n, n)
        return ret

File: 22.generate-parentheses.cpp
class Solution {
public:
    void add(vector<string> &ret, string s, int n, int m) {
        if (n > m) {
            return;
        }
        if (m == 0 && n == 0) {
            ret.push_back(s);
        }
        if (n > 0) {
            add(ret, s + '(', n - 1, m);
        }
        if (m > 0) {
            add(ret, s + ')', n, m - 1);
        }
    }
    vector<string> generateParenthesis(int n) {
        vector<string> ret;
        add(ret, "", n, n);
        return ret;
    }
};